test/src/PublicERC6492Validator/isValidSignatureNow.t.sol

```diff
@@ -0,0 +1,136 @@
+import {CoinbaseSmartWalletFactory} from "smart-wallet/CoinbaseSmartWalletFactory.sol";
+
+contract IsValidSignatureNowTest is SpendPermissionManagerBase {
+    PublicERC6492Validator public validator;
+
+    function setUp() public {
+        _initialize(); // Sets up account, owner, etc
+        mockCoinbaseSmartWalletFactory = new CoinbaseSmartWalletFactory(address(account));
+        validator = new PublicERC6492Validator();
+    }
+
+    function test_isValidSignatureNow_success_preDeployedWallet() public {
+        // Create test message and signature
+        bytes32 hash = keccak256("test message");
+        bytes32 replaySafeHash = account.replaySafeHash(hash);
+        bytes memory signature = _sign(ownerPk, replaySafeHash);
+        bytes memory wrappedSignature = _applySignatureWrapper(0, signature);
+
+        bool isValid = validator.isValidSignatureNow(address(account), hash, wrappedSignature);
+        assertTrue(isValid);
+    }
+
+    // Bug in solady code!
+    // function test_isValidSignatureNow_success_counterfactualWallet() public {
+    //     // Create a new owner for the counterfactual wallet
+    //     uint256 newOwnerPk = uint256(keccak256("different owner"));
+    //     address newOwner = vm.addr(newOwnerPk);
+
+    //     // Setup counterfactual wallet data with new owner
+    //     bytes[] memory owners = new bytes[](1);
+    //     owners[0] = abi.encode(newOwner);
+    //     address counterfactualWallet = mockCoinbaseSmartWalletFactory.getAddress(owners, 0);
+
+    //     // Create test message and signature
+    //     bytes32 hash = keccak256("test message");
+
+    //     // Construct replaySafeHash manually for counterfactual wallet
+    //     bytes32 domainSeparator = keccak256(
+    //         abi.encode(
+    //             keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+    //             keccak256(bytes("Coinbase Smart Wallet")),
+    //             keccak256(bytes("1")),
+    //             block.chainid,
+    //             counterfactualWallet
+    //         )
+    //     );
+    //     bytes32 replaySafeHash =
+    //         keccak256(abi.encodePacked("\x19\x01", domainSeparator, keccak256(abi.encode(CBSW_MESSAGE_TYPEHASH,
+    // hash))));
+
+    //     // Sign with the new owner's key
+    //     bytes memory signature = _sign(newOwnerPk, replaySafeHash);
+    //     bytes memory wrappedSignature = _applySignatureWrapper(0, signature);
+
+    //     // Wrap in ERC-6492 format
+    //     bytes memory eip6492Signature = _wrap6492Signature(wrappedSignature, owners);
+    //     bool isValid = validator.isValidSignatureNow(counterfactualWallet, hash, eip6492Signature); // DOES NOT WORK
+    // (indicates bug in solady code)
+    //     bool isValid = validator.isValidSignatureNowAllowSideEffects(counterfactualWallet, hash, eip6492Signature);
+    // // WORKS as long as it can deploy the wallet
```

Comment:
note lines 64 and 67. 64 does not work, but 67 does

Reply:
(i.e. works when calling the version that allows side effects but fails when calling the non-side-effect version, same signature)

Reply:
not sure what the actual bug is yet

Reply:
I did some investigation into this and I can share my analysis if it's of interest:

I think the reason the signature is failing with `isValidERC6492SignatureNow()` but succeeding with `isValidERC6492SignatureNowAllowSideEffects()` is because `isValidERC6492SignatureNow()` uses the hardcoded address `0x00007bd799e4A591FeA53f8A8a3E9f931626Ba7e`, and this may not exist in the testing environment. For example with the following test function:

```solidity
function test_isValidSignatureNow_success_counterfactualWallet() public {
    // Create a new owner for the counterfactual wallet
    uint256 newOwnerPk = uint256(keccak256("different owner"));
    address newOwner = vm.addr(newOwnerPk);

    // Setup counterfactual wallet data with new owner
    bytes[] memory owners = new bytes[](1);
    owners[0] = abi.encode(newOwner);
    address counterfactualWallet = mockCoinbaseSmartWalletFactory.getAddress(owners, 0);

    // Create test message and signature
    bytes32 hash = keccak256("test message");

    // Construct replaySafeHash manually for counterfactual wallet
    bytes32 domainSeparator = keccak256(
        abi.encode(
            keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
            keccak256(bytes("Coinbase Smart Wallet")),
            keccak256(bytes("1")),
            block.chainid,
            counterfactualWallet
        )
    );
    bytes32 replaySafeHash = keccak256(abi.encodePacked("\x19\x01", domainSeparator, keccak256(abi.encode(CBSW_MESSAGE_TYPEHASH, hash))));

    // Sign with the new owner's key
    bytes memory signature = _sign(newOwnerPk, replaySafeHash);
    bytes memory wrappedSignature = _applySignatureWrapper(0, signature);

    // Wrap in ERC-6492 format
    bytes memory eip6492Signature = _wrap6492Signature(wrappedSignature, owners);
    bool isValid1 = validator.isValidSignatureNow(counterfactualWallet, hash, eip6492Signature);
    bool isValid2 = validator.isValidSignatureNowAllowSideEffects(counterfactualWallet, hash, eip6492Signature);
    assertTrue(isValid1);
    assertTrue(isValid2);
} 
```

I am observing a failure with the command:

```
forge test --match-contract IsValidSignatureNowTest
```

and a success with the command:

```
forge test --match-contract IsValidSignatureNowTest --rpc-url <INSERT_ETH_RPC_URL>
```

In the future if there's a scenario where `isValidERC6492SignatureNow()` needs to be tested without using `--rpc-url`, one idea is to add the factory call that's described in [this Solady GitHub gist](https://gist.github.com/Vectorized/846a474c855eee9e441506676800a9ad):


```diff
function test_isValidSignatureNow_success_counterfactualWallet() public {
    // ...

+   address FACTORY = 0x4e59b44847b379578588920cA78FbF26c0B4956C;
+   FACTORY.call(hex"000000000000000000000000000000000000000068f35e1510740001fd13984a6040600b3d3960403df3fe36383d373d3d6020515160208051013d3d515af160203851516084018038385101606037303452813582523838523490601c34355afa34513060e01b141634fd");
    
    bool isValid1 = validator.isValidSignatureNow(counterfactualWallet, hash, eip6492Signature);
    bool isValid2 = validator.isValidSignatureNowAllowSideEffects(counterfactualWallet, hash, eip6492Signature);
    assertTrue(isValid1);
    assertTrue(isValid2);
}
```